nes-logical-operators/src/Plans/LogicalPlanBuilder.cpp

```diff
@@ -197,9 +197,8 @@ LogicalPlanBuilder::checkAndAddWatermarkAssigner(LogicalPlan queryPlan, const st
 }
 
 LogicalPlan LogicalPlanBuilder::addBinaryOperatorAndUpdateSource(
-    const LogicalOperator& operatorNode, LogicalPlan leftLogicalPlan, LogicalPlan rightLogicalPlan)
+    const LogicalOperator& operatorNode, const LogicalPlan& leftLogicalPlan, const LogicalPlan& rightLogicalPlan)
 {
-    leftLogicalPlan.rootOperators.push_back(rightLogicalPlan.rootOperators[0]);
-    return promoteOperatorToRoot(leftLogicalPlan, operatorNode);
+    return promoteOperatorToRoot(leftLogicalPlan.withAdditionalRoots(rightLogicalPlan.getRootOperators()), operatorNode);
```

Comment:
concatenation can be done here with a bit of ranges codes

```suggestion
auto newRootOperators = std::ranges::views::join(std::array{leftLogicalPlan.getRootOperators(), rightLogicalPlan.getRootOperators()});
+    return promoteOperatorToRoot(leftLogicalPlan.withRootOperators(newRootOperators | std::ranges::to<std::vector>()), operatorNode);
```